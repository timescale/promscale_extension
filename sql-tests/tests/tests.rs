use insta::assert_snapshot;
use std::{
    env,
    mem::MaybeUninit,
    sync::{Mutex, Once},
};
use test_common::*;
use test_generator::test_resources;

const TESTDATA: &'static str = concat!(env!("CARGO_MANIFEST_DIR"), "/testdata");

fn init_lock() -> &'static Mutex<()> {
    static mut SINGLETON: MaybeUninit<Mutex<()>> = MaybeUninit::uninit();
    static ONCE: Once = Once::new();

    unsafe {
        ONCE.call_once(|| {
            let singleton = Mutex::new(());
            SINGLETON.write(singleton);
        });

        SINGLETON.assume_init_ref()
    }
}

// TODO fix how test_resources works in nexted workspaces
#[test_resources("sql-tests/testdata/*.sql")]
fn sql_tests(full_resource: &str) {
    let pg_blueprint = PostgresContainerBlueprint::new().with_testdata(TESTDATA);
    let test_pg_instance = new_test_instance_from_env(&pg_blueprint);
    let mut init_conn = test_pg_instance.connect();
    // This lock prevents multiple callers from running
    // CREATE EXTENSION simultaneously. In case of
    // promscale_extension concurrent CREATE EXTENSION causes
    // issues when migration incremental/003-users.sql executes.
    {
        let _init_lock = init_lock().lock().unwrap();
        init_conn
            .simple_query("CREATE EXTENSION promscale;")
            .expect("Unable to create extension promscale.");
    }

    let resource = if let Some((_, rest)) = full_resource.split_once('/') {
        rest
    } else {
        full_resource
    };
    let query_result = test_pg_instance.exec_sql_script(resource);
    assert_snapshot!(resource, query_result);
}
