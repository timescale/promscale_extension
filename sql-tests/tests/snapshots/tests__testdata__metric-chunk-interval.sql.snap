---
source: sql-tests/tests/tests.rs
expression: query_result
---
 plan 
------
 1..4
(1 row)

                       is                        
-------------------------------------------------
 ok 1 - default metric chunk interval is 8 hours
(1 row)

 set_default_chunk_interval 
----------------------------
 t
(1 row)

                       is                       
------------------------------------------------
 ok 2 - default metric chunk interval is 1 hour
(1 row)

 get_or_create_metric_table_name 
---------------------------------
 (1,cpu_usage,t)
(1 row)

                               is                                
-----------------------------------------------------------------
 ok 3 - get_metric_chunk_interval returns default chunk interval
(1 row)

 set_metric_chunk_interval 
---------------------------
 t
(1 row)

                        approx_is                        
---------------------------------------------------------
 ok 4 - get_metric_chunk_interval returns chunk interval
(1 row)

 finish 
--------
(0 rows)


